### PR TITLE
Fcos num pos

### PR DESCRIPTION
**Pull request by yhcao6**
| Model             | 4GPU | 8GPU |
|-------------------|------|------|
| R50-GN-1x         | 36.6 | 36.7 |
| R50-GN-2x         | 36.9 | 36.2 |
| R50-GN-mstrain-2x | 38.7 | 38.5 |
| R101-GN-1x | 39.2 | 39.1|
| R101-GN-2x | 39.1 | 38.7|
| R101-GN-mstrain-2x | 40.9 | 40.8 |
| X101-GN-mstrain-2x | 42.5 | 42.6 |
| R50-tricks| 38.6|38.7 |
| R50-tricks-dcn-mstrain-2x| 42.5|42.3|

### CODE REVIEW

**Comment by hiyyg** [comment]
Why is FCOS in mmdet much worse than that of adet (https://github.com/aim-uofa/AdelaiDet/tree/master/configs/FCOS-Detection)?

**Comment by yhcao6** [comment]
> Why is FCOS in mmdet much worse than that of adet (https://github.com/aim-uofa/AdelaiDet/tree/master/configs/FCOS-Detection)?

We don't use all tricks.

**Comment by hiyyg** [comment]
why not?

**Comment by yhcao6** [comment]
> why not?

We have implemented these tricks, you can refer to https://github.com/open-mmlab/mmdetection/blob/master/configs/fcos/fcos_center-normbbox-centeronreg-giou_r50_caffe_fpn_gn-head_4x4_1x_coco.py.
The baseline config should be consistent with the precision in the original paper. https://arxiv.org/abs/1904.01355
This Pr aims to supports training with 8 GPU without the precision drop. I just picked several models for testing.

File: mmdet/models/dense_heads/fcos_head.py [hunk 1]
```diff
@@ -219,6 +219,7 @@ def loss(self,
         pos_inds = ((flatten_labels >= 0)
                     & (flatten_labels < bg_class_ind)).nonzero().reshape(-1)
         num_pos = len(pos_inds)
+        num_pos = max(reduce_mean(torch.tensor(num_pos).cuda()).item(), 1.0)
```
Comment (shinya7y) [inline mmdet/models/dense_heads/fcos_head.py:222]: Casting like the below may be needed for PyTorch 1.6+. https://github.com/open-mmlab/mmdetection/pull/4210
``` python
num_pos = torch.tensor(num_pos, dtype=torch.float, device=bbox_preds[0].device)
num_pos = max(reduce_mean(num_pos).item(), 1.0)
```

File: mmdet/models/dense_heads/fcos_head.py [hunk 2]
```diff
@@ -219,6 +219,7 @@ def loss(self,
...
         loss_cls = self.loss_cls(
             flatten_cls_scores, flatten_labels,
             avg_factor=num_pos + num_imgs)  # avoid num_pos is 0
```
Comment (shinya7y) [inline mmdet/models/dense_heads/fcos_head.py:225]: `# avoid num_pos is 0` becomes inconsistent with the code.
Is ` + num_imgs` still needed for AP?

**Review by shinya7y — CHANGES_REQUESTED** [review]
https://github.com/open-mmlab/mmdetection/blob/cd00b884759a2572021461e7e8204f364581e01d/mmdet/models/dense_heads/fcos_head.py#L230
``` python
if len(pos_inds) > 0:
```

**Review by shinya7y — COMMENTED** [review]
Many dependencies in the HRNet directory.
https://github.com/open-mmlab/mmdetection/tree/cd00b884759a2572021461e7e8204f364581e01d/configs/hrnet
https://github.com/open-mmlab/mmdetection/tree/cd00b884759a2572021461e7e8204f364581e01d/configs/hrnet/fcos_hrnetv2p_w32_gn-head_4x4_1x_coco.py

Updating config links and Notes in the FCOS README is also needed.

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**